core/matrix/dense.cpp

```diff
@@ -1152,90 +1292,138 @@ void Dense<ValueType>::row_gather_impl(const Dense<ValueType>* alpha,
+    auto result = Dense::create(this->get_executor(), this->get_size());
+    this->permute(permutation_indices, result);
+    return result;
 }
 
 
 template <typename ValueType>
-template <typename IndexType>
-void Dense<ValueType>::inverse_column_permute_impl(
-    const array<IndexType>* permutation_indices, Dense<ValueType>* output) const
+std::unique_ptr<Dense<ValueType>> Dense<ValueType>::permute(
+    ptr_param<const Permutation<int32>> permutation, permute_mode mode) const
 {
-    GKO_ASSERT_EQ(permutation_indices->get_num_elems(), this->get_size()[1]);
-    GKO_ASSERT_EQUAL_DIMENSIONS(this, output);
-    auto exec = this->get_executor();
+    auto result = Dense::create(this->get_executor(), this->get_size());
+    this->permute(permutation, result, mode);
+    return result;
+}
+
 
-    exec->run(dense::make_inverse_column_permute(
-        make_temporary_clone(exec, permutation_indices).get(), this,
-        make_temporary_output_clone(exec, output).get()));
+template <typename ValueType>
+std::unique_ptr<Dense<ValueType>> Dense<ValueType>::permute(
+    ptr_param<const Permutation<int64>> permutation, permute_mode mode) const
+{
+    auto result = Dense::create(this->get_executor(), this->get_size());
+    this->permute(permutation, result, mode);
+    return result;
 }
 
 
 template <typename ValueType>
-std::unique_ptr<LinOp> Dense<ValueType>::permute(
-    const array<int32>* permutation_indices) const
+std::unique_ptr<Dense<ValueType>> Dense<ValueType>::permute(
+    ptr_param<const Permutation<int32>> row_permutation,
+    ptr_param<const Permutation<int32>> col_permutation, bool invert) const
 {
     auto result = Dense::create(this->get_executor(), this->get_size());
-    this->permute(permutation_indices, result);
+    this->permute(row_permutation, col_permutation, result, invert);
     return result;
 }
 
 
 template <typename ValueType>
-std::unique_ptr<LinOp> Dense<ValueType>::permute(
-    const array<int64>* permutation_indices) const
+std::unique_ptr<Dense<ValueType>> Dense<ValueType>::permute(
+    ptr_param<const Permutation<int64>> row_permutation,
+    ptr_param<const Permutation<int64>> col_permutation, bool invert) const
 {
     auto result = Dense::create(this->get_executor(), this->get_size());
-    this->permute(permutation_indices, result);
+    this->permute(row_permutation, col_permutation, result, invert);
     return result;
 }
 
 
+template <typename ValueType>
+void Dense<ValueType>::permute(ptr_param<const Permutation<int32>> permutation,
+                               ptr_param<Dense<ValueType>> result,
+                               permute_mode mode) const
+{
+    this->permute_impl(permutation.get(), mode, result.get());
+}
+
+
+template <typename ValueType>
+void Dense<ValueType>::permute(ptr_param<const Permutation<int64>> permutation,
+                               ptr_param<Dense<ValueType>> result,
+                               permute_mode mode) const
```

Comment:
I think it is a new interface, we can make the ordering like `permute_impl` move the output to the end?

Reply:
can you please elaborate? I don't understand what you mean by moving the output.

Reply:
internal one: `permute_impl(perm, mode, result)` the result is the last one arguments.
public: `permute(perm, result, mode)` the result is in the middle.
I think it can also be `permute(perm, mode, result)` if it does not introduce interface break.